aten/src/ATen/core/op_registration/kernel_functor.h

```diff
@@ -0,0 +1,183 @@
+      KernelFunctor* functor = static_cast<KernelFunctor*>(cache);
+      auto output = call_functor_with_ivalue_args<KernelFunctor>(functor, torch::jit::last(*stack, num_inputs));
+      torch::jit::pop(*stack, num_inputs);
+      push_outputs<typename guts::infer_function_traits_t<KernelFunctor>::return_type>::call(std::move(output), stack);
+    }
+  };
+
+  // SFINAE version for kernels that don't return an output
+  template<class KernelFunctor>
+  struct wrap_kernel_functor<KernelFunctor, guts::enable_if_t<std::is_same<void, typename guts::infer_function_traits_t<KernelFunctor>::return_type>::value>> final {
+    static void call(Stack* stack, KernelCache* cache) {
+      static_assert(std::is_base_of<OperatorKernel, KernelFunctor>::value, "Kernel functor must inherit from c10::OperatorKernel");
+
+      constexpr size_t num_inputs = guts::infer_function_traits_t<KernelFunctor>::number_of_parameters;
+      KernelFunctor* functor = static_cast<KernelFunctor*>(cache);
+      call_functor_with_ivalue_args<KernelFunctor>(functor, torch::jit::last(*stack, num_inputs));
+      torch::jit::pop(*stack, num_inputs);
+    }
+  };
+
+  template<class KernelFunctor, class... Args>
+  class KernelFactory final {
+  public:
+    explicit constexpr KernelFactory(Args... args)
+    : constructor_parameters_(std::move(args)...) {}
+
+    std::unique_ptr<KernelCache> operator()() const {
+      return guts::apply(
+        [] (const Args&... params) {return guts::make_unique<KernelFunctor>(params...); },
+        constructor_parameters_);
+    }
+
+  private:
+    std::tuple<Args...> constructor_parameters_;
+  };
+}
+
+/**
+ * Use this to register an operator whose kernel is implemented as a functor
+ *
+ * Example:
+ *
+ * > namespace {
+ * >   class my_kernel_cpu final : public c10::OperatorKernel {
+ * >   public:
+ * >     Tensor operator()(Tensor a, Tensor b) {...}
+ * >   };
+ * > }
+ * >
+ * > static auto registry = c10::RegisterOperators()
+ * >     .op("my_op",
+ * >         c10::kernel<my_kernel_cpu>(),
+ * >         c10::dispatchKey(CPUTensorId()));
+ *
+ * The functor constructor can take arguments to configure the kernel.
+ * The arguments are defined in the kernel registration.
+ * Example:
+ *
+ * > namespace {
+ * >   class my_kernel_cpu final : public c10::OperatorKernel {
+ * >   public:
+ * >     explicit my_kernel_cpu(std::string some_configuration, int a, bool b)
+ * >         : ... {...}
+ * >
+ * >     Tensor operator()(Tensor a, Tensor b) {...}
+ * >   };
+ * > }
+ * >
+ * > static auto registry = c10::RegisterOperators()
+ * >     .op("my_op",
+ * >         c10::kernel<my_kernel_cpu>("some_configuration", 3, true),
+ * >         c10::dispatchKey(CPUTensorId()));
+ */
+template<class KernelFunctor, class... ConstructorParameters>
+inline constexpr auto kernel(ConstructorParameters&&... constructorParameters)
+// enable_if: only enable it if KernelFunctor is actually a functor and inherits from c10::OperatorKernel
```

Comment:
What does the error message look like when these conditions are not met?

Reply:
Bad. I assume it would say it doesn't find the correct overload for `kernel()`. I'm planning to have a separate PR up top that improves error messages.